File: 206-ReverseLinkedList/206-ReverseLinkedList.py
# Last updated: 6/28/2026, 9:06:33 PM
1# Definition for singly-linked list.
2# class ListNode:
3#     def __init__(self, val=0, next=None):
4#         self.val = val
5#         self.next = next
6class Solution:
7    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
8        if not head:
9            return head
10        prev = head
11        curr = head.next
12        prev.next = None
13        while curr:
14            temp = curr.next
15            curr.next = prev
16            prev = curr
17            curr = temp
18        return prev

File: 206-ReverseLinkedList/206-ReverseLinkedList.java
// Last updated: 8/16/2026, 9:51:26 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    // public ListNode reverseList(ListNode head) {
    //     ListNode newHead = null; 

    //     while(head != null) {
    //         ListNode next = head.next; // next=2
    //         head.next = newHead; // set 1->null (newHead)
            
    //         newHead = head; // update newHead to head = 1
    //         head = next; // move to next node, from 1 to 2
    //     }
    //     return newHead;
    // }

    // method 2 - recursion - Traversal
    // recursion = {Traversal [top-down], Divide and Conquer [bottom-up]}
    public ListNode reverseList(ListNode head) {
        return helper(head, null);
        // return reversed Listnode with newHead=null
    }

    private ListNode helper(ListNode head, ListNode newHead) {
        // return reversed ListNode with head replaced to newHead
        if(head == null) return newHead; //recursion exit

        ListNode next = head.next;
        head.next = newHead;

        return helper(next, head);

    }
}